operator/src/main/resources/scripts/stop-server.py

```diff
@@ -2,23 +2,197 @@
+# NOTE: The overall timeout used to control the termination of this container is
+# controlled by the Kubernetes graceful timeout value. So even
+# if this code looped forever, Kubernetes will kill the pod once the timeout expires.
+# The user must set that timeout to a large enough value to give Coherence time to get safe.
+def waitUntilCoherenceSafe():
+  print ('Shutdown: getting all service Coherence MBeans')
+
+  domainRuntime()
+  query='Coherence:type=Service,name=*,*'
+
+  # Wait forever until we get positive ack that it is ok to shutdown this server.
+  done = False
+  while (not done):
+    try:
+      beans = list(mbs.queryMBeans(ObjectName(query), None))
+      print("Shutdown: Coherence service bean count " + str(len(beans)))
+
+      # Loop waiting for each service to be safe
+      for serviceBean in beans:
+        objectName = serviceBean.getObjectName()
+        waitUntilServiceSafeToShutdown(objectName)
+      done = True
+      print ('Shutdown: It is safe to shutdown Coherence')
+
+    except:
+      print ("Shutdown: Exception checking a service Coherence statusHA, retrying...")
+      traceback.print_exc(file=sys.stdout)
+      dumpStack()
+      systime.sleep(30)
+      pass
+
+
+# Wait until the specified Coherence service is safe to shutdown.
+# If the cluster is a single node cluster then the service will always
+# be ENDANGERED, therefore it is the responsibility of the user to
+# set Coherence backup count to 0, or to set the terminate grace period
+# to a low number since this method will just wait until the oeprator kills the
+# pod.
+def waitUntilServiceSafeToShutdown(objectName):
+
+  print ("Shutdown: checking Coherence service " + str(objectName) )
+
+  # NOTE: break loop when it safe to shutdown else stay in loop forever
+  while (True):
+    try:
+      # If the BackupCount is > 0 then the user intention is to have
+      # HA with backed up partitions, otherwise if value is < 1 then the user
+      # doesn't care about HA so no need to wait.
+      # NOTE: if this is NOT a partitioned service we will get exception, ignore and exit loop
+      try:
+        val = mbs.getAttribute(objectName,"BackupCount")
+      except:
+        print ("Shutdown: Coherence BackupCount attribute missing")
+        val = None
+        pass
+
+      if (val is None) or int(val) < 1:
+        print ("Shutdown: Coherence skipping status check for this service since BackupCount < 1.  Value is " + str(val))
+        break
+
+      status = mbs.getAttribute(objectName,"StatusHA")
+      if (status is None):
+        print ("Shutdown: None returned for Coherence StatusHA")
+        break
+
+      print ('Shutdown: Coherence StatusHA is ' + status)
+      if status != "ENDANGERED":
+        break
+
+      # Coherence caches are ENDANGERED meaning that we may lose data
+      print ('Shutdown: Waiting until it is safe to shutdown Coherence server ...')
+      systime.sleep(5)
```

Comment:
Maybe just sleep 1 second?  I doubt the overhead of the mbean call is significant enough that it matters if we call it more often.  That'd speedup shutdown...     (If you do bring this down to 1 second please add code at line 181 to so that the 'Waiting until it is safe...' message is only printed once instead of once per second.)

Reply:
fixed